Feature/jwt authentication

ChawinTan opened the pull request:
<!--
For support requests, please read the Support Guidelines to know where to ask: https://github.com/open-rmf/rmf/wiki/Support-guidelines
For general questions and design discussion, please use the Discussions page: https://github.com/open-rmf/rmf/discussions
Not sure if this is the right repository? Open an issue on https://github.com/open-rmf/rmf
We require contributors to GPG Sign their commits. Follow the guide here to set up a GPG key and add it to your GitHub account: https://docs.github.com/en/github/authenticating-to-github/generating-a-new-gpg-key
A quick guide to how to sign your commits can be seen here: https://gist.github.com/mort3za/ad545d47dd2b54970c102fe39912f305
To GPG sign several commits at once (for example, if you forgot to sign some commits), you can run this command, followed by a force-push: git rebase --exec 'git commit --amend --no-edit -n -S' -i <commit before the first commit you want signed>
For bug fix pull requests, please fill out the information below.
Be as detailed as possible.
-->

### Implemented feature

***Refer to [this pull request](https://github.com/open-rmf/rmf_visualization/pull/15) for the actual pr***

<!-- Briefly describe the feature being implemented.
If there is a feature request issue for the feature, link to that feature.
If there is not a feature request issue for the feature, create one first and fill out all the required information there, then link to that issue from this new feature pull request. -->

Feature Request: https://github.com/open-rmf/rmf-web/issues/277

Below is the library that I intend to use:
https://github.com/Thalhammer/jwt-cpp

This feature is to allow schedule visualizer's websocket server to verify jwt tokens from an external user and accept the connections only if the tokens are valid when in use with rmf-web dashboard.

### Implementation description

<!-- Describe the approach taken to implement the feature.
Provide a link to a detailed design document and discussion of that design.
Implementations without design documentation will not be accepted until design documentation has been provided and discussed.
Usually this is done via the feature request issue. -->

**For reviewing purposes**
1) follow the instructions [here](https://github.com/open-rmf/rmf) to download the source code
2) `cd src/rmf/rmf_visualization`
3) check out this branch for rmf_visualizer. `git checkout -b feature/jwt-authentication origin/feature/jwt-authentication` 
4) `cd ~/rmf_ws`
5) `source /opt/ros/foxy/setup.bash` and then `colcon build --cmake-args -DCMAKE_BUILD_TYPE=Release`

**For starting rmf_web**
Refer to the instructions [here](https://github.com/open-rmf/rmf-web) to start the dashboard.

Add a `token` parameter to `TrajectoryRequest` and `NegotiationTrajectoryRequest` in rmf-web.

When running with `rmf-web`, you should source an environment variable `export JWT_PUBLIC_KEY=<your token>`.

**Description**
Within the `on_message` method, the first step is to check for the presence of `JWT_PUBLIC_KEY` environment variable that will be provided when running with the dashboard.

If `JWT_PUBLIC_KEY` exist, then if will proceed to get and verify the token from the message payload. If no token is provided or token is invalid, an error message will be shown on the console and a message payload with an error message will be sent to the client.

The accessibility would be controlled by the `is_verified` flag. If it is false, then the server would not send the request for processing. 

The library is added via git submodules during colcon build.

An example of a connection refused when no token/token is invalid:
![token_error](https://user-images.githubusercontent.com/29136440/114349213-30260900-9b9a-11eb-8655-cf80b1730a3b.png)

koonpeng reviewed (changes requested):
` if (url) this._backendWs = new WebSocket(url + '/?token=<your_token>')`
Do not put the token in the url, this is a security risk. See https://tools.ietf.org/html/rfc6750#page-6

```
   Because of the security weaknesses associated with the URI method
   (see Section 5), including the high likelihood that the URL
   containing the access token will be logged, it SHOULD NOT be used
   unless it is impossible to transport the access token in the
   "Authorization" request header field or the HTTP request entity-body.
   Resource servers MAY support this method.

   This method is included to document current use; its use is not
   recommended, due to its security deficiencies (see Section 5) and
   also because it uses a reserved query parameter name, which is
   counter to URI namespace best practices, per "Architecture of the
   World Wide Web, Volume One" [W3C.REC-webarch-20041215].
```

You choices are
1. Authorization header, can't use this because we can't send custom headers in a websocket request.
2. Cookies, beware of CSRF, ~to keep it safe, the cookie needs to be http-only. Since we dont control the auth server we can't exactly set this flag so the risk of CSRF is always present.~ http-only is to avoid XSS, avoiding CSRF is more complicated, one of the ways is to use a ephermal token, but we can't exactly do it with a websocket-only server.
3. Application level. Basically verify the token as part of the app, meaning to accept all connection, but lock all functionalities until the client sends a valid token.
4. Nonstandard websocket protocol. While you cannot set custom headers, one of the header you can set is the websocket protocol header ("Sec-protocol" or something similar). You can put the token in that header and the server will check it there.
5. Finally, the end-all solution would be to make a full-fledge oauth2 client. This means that you will register the app server to keycloak as a oauth2 client, the token is never exposed to the user agent, the app server will obtain the access token from the auth provider using the client_id and client_secret. The biggest challenge of this is that you need to have a session store, and possibly a database to persist said sessions, this is also probably the only way to properly implement option 2.

IMO option 3 or 4 would be the easiest.

ChawinTan commented:
Thanks for all the suggestions! 

For option 3, does that mean that the token would be send in the message payload itself and every request would have to be verified, similar to how the "role base" unlocking of features would work?

ChawinTan commented:
Realized that throwing an exception that causes the process to shut down is a bad idea. I introduced a boolean flag `is_verified` that controls the availability of the services depending on the verification outcome.

Refactored receive the jwt tokens on message payload rather than on connection.

Since we are passing the tokens in the message payload, how should we let the user know that the service is not available since the connection will still be going through? Should we create and send a message that informs them of the bad token or how is it normally done in such situations?

koonpeng commented:
> For option 3, does that mean that the token would be send in the message payload itself and every request would have to be verified, similar to how the "role base" unlocking of features would work?

Verifying per request is more secure, but I think you can look at socket.io and see how they do it. I think they only verify once on connection, I think this is more inline with how websockets typically works.

ChawinTan commented:
> Note that for verifying per request, there is a "bug" currently in the client in that it doesn't periodically refresh the token, so eventually you will get expired tokens.

What do you think of sending an error payload like the one below and when the client finds out that a token is expired, generates a new one? The error message is copied from the exception that is caught from the verifier. We can also get even more specific with the error messages when checking with parameters, eg `.with_issuer("auth0");`
![token_expired](https://user-images.githubusercontent.com/29136440/114896606-122f0180-9e43-11eb-9d69-b12fa3798b16.png)

I think this can be one of the TODOs as well, along with unifying the 2 separate websockets on the client and refactoring the client to handle cases where an error message is passed back.

For now though, the websocket server seems to be doing what I want it to do.

koonpeng commented:
The client knows when the token will expire so it can proactively refresh it without input from the server. The keycloak client has a function to refresh tokens, but I don't know if it is the access token or id token (or both). They may have different expriation, we are using the id token so we need to make sure that gets refreshed as well.

ChawinTan commented:
Design doc for the flow of the program:
![Jwt_authentication_design_doc](https://user-images.githubusercontent.com/29136440/115489760-5d0ba780-a28f-11eb-961f-04373756335f.png)

koonpeng commented on .gitmodules:
```diff
@@ -0,0 +1,3 @@
+[submodule "rmf_visualization_schedule/jwt-cpp"]
```
looks like this points to a commit in master branch, I would recommend to use a release tag instead.

koonpeng reviewed (dismissed):
No objections with the code, but I would wait for the accompanying PR in rmf-web (refresh token pr) before merging this.

koonpeng commented:
> No objections with the code, but I would wait for the accompanying PR in rmf-web (refresh token pr) before merging this.

On 2nd thought, it doesn't matter as auth is disabled by default.

On a similar topic, does this support https or is there an example deployment enabling https? Because auth is useless without a secure connection.

ChawinTan commented:
> On a similar topic, does this support https or is there an example deployment enabling https? Because auth is useless without a secure connection.

Not that I know of. I don't think it supports TLS right now. Base on [this](https://github.com/zaphoyd/websocketpp/blob/0.6.0-permessage-deflate/examples/echo_server_tls/echo_server_tls.cpp) example, it seems like we need to use `asio_tls` instead of `asio` and potentially supporting both types.
